### PR TITLE
Implementation of einsum

### PR DESCRIPTION
**Pull request by abergeron**
This doesn't support diag/traces, but should support everything else.  If we ever add support for diag, then I can probably add support for them in the macro.

There is also a small fix to torch.max to support an extra argument pattern.

### CODE REVIEW

File: tests/operations/test_macro_einsum.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import pytest
+
+from myia.lib import InferenceError
+from myia.operations import einsum
+
+from ..common import MA, MB
+from ..multitest import infer, mt, run_debug
+
+
+@run_debug(MA(4, 1)[0])
```
Comment (breuleux) [inline tests/operations/test_macro_einsum.py]: `MA(4, 1)[0]` returns a vector of size 1, did you mean `MA(1, 4)[0]`? (Same comment for the tests below)

File: tests/operations/test_macro_einsum.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    return einsum('ij->j', a)
+
+
+@run_debug(MA(3, 4))
+def test_einsum_sum2d_01(a):
+    return einsum('ij->', a)
+
+
+@run_debug(MA(3, 4), MB(3, 4))
+def test_einsum_elemwise2d(a, b):
```
Comment (breuleux) [inline tests/operations/test_macro_einsum.py]: This is not an elemwise operation.

File: myia/operations/macro_einsum.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
+"""Implementation of the 'einsum' macro."""
+
+from opt_einsum import contract, contract_expression
+
+from ..lib import InferenceError, macro
+from . import primitives as P
+
+
+def _tensordot(g, a, b, *, axes):
```
Comment (breuleux) [inline myia/operations/macro_einsum.py]: Could you write some light documentation for this and the other private functions in this module? It'd help a little for review/maintenance.

File: tests/operations/test_macro_einsum.py (hunk 1 above)
Comment (abergeron) [reply tests/operations/test_macro_einsum.py]: Yes

File: tests/operations/test_macro_einsum.py (hunk 2 above)
Comment (abergeron) [reply tests/operations/test_macro_einsum.py]: Yes, that should be ij on output.

File: myia/operations/macro_einsum.py (hunk 3 above)
Comment (abergeron) [reply myia/operations/macro_einsum.py]: Done

**Pull request merged**